inference-engine/src/mkldnn_plugin/nodes/mkldnn_one_hot_node.cpp

```diff
@@ -56,30 +47,77 @@ MKLDNNOneHotNode::MKLDNNOneHotNode(const std::shared_ptr<ngraph::Node>& op, cons
     errorPrefix = "OneHot layer with name '" + op->get_friendly_name() + "'";
     const auto oneHot = std::dynamic_pointer_cast<const ngraph::opset1::OneHot>(op);
     const auto depthNode = std::dynamic_pointer_cast<const ngraph::opset1::Constant>(oneHot->get_input_node_shared_ptr(DEPTH_ID));
-    const auto onValueNode = std::dynamic_pointer_cast<const ngraph::opset1::Constant>(oneHot->get_input_node_shared_ptr(ON_VALUE_ID));
-    const auto offValueNode = std::dynamic_pointer_cast<const ngraph::opset1::Constant>(oneHot->get_input_node_shared_ptr(OFF_VALUEAXES_ID));
-    depth = depthNode->cast_vector<uint32_t>()[0];
+    if (depthNode) {
+        depth = depthNode->cast_vector<uint32_t>()[0];
+    }
     axis = oneHot->get_axis();
-    src_dims = oneHot->get_input_shape(INDICES_ID);
-    if (ngraph::is_scalar(src_dims)) {
-        src_dims = SizeVector{1};
+
+    VectorDims srcDims = getInputShapeAtPort(INDICES_ID).getDims();
+    if (ngraph::is_scalar(srcDims)) {
+        srcDims = SizeVector{1};
     }
-    dst_dims = oneHot->get_output_shape(0);
-    if (ngraph::is_scalar(dst_dims)) {
-        dst_dims = SizeVector{1};
+    VectorDims dstDims = getOutputShapeAtPort(0).getDims();
+    if (ngraph::is_scalar(dstDims)) {
+        dstDims = SizeVector{1};
     }
 
-    int output_dims_size = dst_dims.size();
+    int output_dims_size = dstDims.size();
     if (axis < 0) {
         axis += output_dims_size;
     }
     if (axis < 0 || axis >= output_dims_size) {
         IE_THROW() << errorPrefix << " has unsupported 'axis' attribute: " << oneHot->get_axis();
     }
 
-    if (!( ((1 + src_dims.size()) == dst_dims.size()) ||
-           (src_dims.size() == 1 && dst_dims.size() == 1 && dst_dims[0] == depth && src_dims[0] == 1)))
+    if (!(((1 + srcDims.size()) == dstDims.size()) ||
+            (depthNode && (srcDims.size() == 1 && dstDims.size() == 1 && dstDims[0] == depth && srcDims[0] == 1))))
+        IE_THROW() << errorPrefix << " has incorrect number of input/output dimensions!";
+}
+
+bool MKLDNNOneHotNode::needShapeInfer() const {
+    const auto depthNodePtr = reinterpret_cast<int32_t *>(getParentEdgesAtPort(1)[0]->getMemoryPtr()->GetPtr());
+    if (depth != depthNodePtr[0])
+        return true;
+    return MKLDNNNode::needShapeInfer();
+}
+
+std::vector<VectorDims> MKLDNNOneHotNode::shapeInfer() const {
+    ngraph::OutputVector inputsForShapeInfer;
```

Comment:
I propose to use [this helper](https://github.com/openvinotoolkit/openvino/blob/master/inference-engine/src/mkldnn_plugin/utils/shape_inference/shape_inference.cpp#L18) -- in the current state it will handle shape inference through _validate_and_infer_type_ -- just the way you did but the code here would look cleaner

Here is an [example](https://github.com/openvinotoolkit/openvino/blob/master/inference-engine/src/mkldnn_plugin/nodes/mkldnn_reshape_node.cpp#L94) how it is done for Reshape operation. 
Please take a look how constant data being communicated via parameter of _shape_inference_ helper

In the future we will update this helper for faster shape inference (no validate_and_infer_types will be used). With _shape_inference_ helper OneHot would pick up these changes instantly with no _MKLDNNOneHotNode::shapeInfer()_ changes.